### PR TITLE
Added StatesCollector class

### PR DESCRIPTION
**Pull request by chrisdembia**
This is a new-style reporter that accumulates states during a simulation. The Collector holds onto a StatesTrajectory that someone could access at the end of a simulation.

The StatesTrajectory within the collector is cleared when realizing to Instance.

I chose the name StatesCollector instead of StatesReporter because we already have a StatesReporter analysis.

### CODE REVIEW

File: OpenSim/Simulation/StatesCollector.cpp [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+    m_states.clear();
+}
+
+const StatesTrajectory& StatesCollector::getStates() const {
+    return m_states;
+}
+
+void StatesCollector::extendRealizeInstance(const SimTK::State& state) const {
+    Super::extendRealizeInstance(state);
+    const_cast<StatesCollector*>(this)->clear();
```
Comment (tkuchida) [inline OpenSim/Simulation/StatesCollector.cpp]: Is this the intended use of `extendRealizeInstance()`? If so, why is it a const method? :dromedary_camel:

File: OpenSim/Simulation/StatesCollector.h [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+/** Stores the states during a simulation.
+ *
+ * This class was introduced in v4.0 and is intended to replace the
+ * StatesReporter analysis.
+ */
+class OSIMSIMULATION_API StatesCollector : public ModelComponent {
+OpenSim_DECLARE_CONCRETE_OBJECT(StatesCollector, ModelComponent);
+
+public:
+    /** Access the recorded states. */
```
Comment (tkuchida) [inline OpenSim/Simulation/StatesCollector.h]: Perhaps "recorded" -> "collected" for consistency, or change StatesCollector to StatesRecorder?

File: OpenSim/Simulation/StatesCollector.cpp (hunk 1 above)
Comment (chrisdembia) [inline OpenSim/Simulation/StatesCollector.cpp]: I'm not exactly sure. @aseth1 ?

File: OpenSim/Simulation/StatesCollector.h (hunk 2 above)
Comment (tkuchida) [inline OpenSim/Simulation/StatesCollector.h]: Actually, StatesAccumulator seems like a more accurate description IMHO. :hamburger:

File: OpenSim/Simulation/Test/testStatesTrajectory.cpp [hunk 3]
```diff
@@ -54,12 +54,17 @@ Real getStorageEntry(const Storage& sto,
...
+void testPopulateTrajectoryAndStatesCollector() {
     Model model("gait2354_simbody.osim");
 
     // To assist with creating interesting (non-zero) coordinate values:
     model.updCoordinateSet().get("pelvis_ty").setDefaultLocked(true);
 
+    // Also, test the StatesCollector.
+    auto* statesCol = new StatesCollector();
+    statesCol->setName("states_collector");
+    model.addModelComponent(statesCol);
```
Comment (tkuchida) [inline OpenSim/Simulation/Test/testStatesTrajectory.cpp]: Adding `statesCol` to the Model doesn't quite seem right to me. Will it eventually be added to a Study?

**Comment by tkuchida** [comment]
`</review>` Everything else looks good to me :dolphin:

File: OpenSim/Simulation/Test/testStatesTrajectory.cpp (hunk 3 above)
Comment (chrisdembia) [inline OpenSim/Simulation/Test/testStatesTrajectory.cpp]: Yes.

**Pull request closed without merging**